payu/models/staged_cable.py

```diff
@@ -0,0 +1,205 @@
+                for step_name, step_opts in stage_opts.items():
+                    step_names.append(step_name)
+                    step_counts.append(step_opts['count'])
+
+                # Now iterate to the maximum number of steps
+                for step in range(max(step_counts)):
+                    for step_id, _ in enumerate(step_names):
+                        if step_counts[step_id] > step:
+                            cable_stages.append(step_names[step_id])
+            # Finish handling of multistep stage
+
+            else:
+                # A single step stage, in general we only want to run this
+                # once, but check for the count anyway
+                for _ in range(stage_opts['count']):
+                    cable_stages.append(stage_name)
+
+        # Finish handling of single step stage
+        return cable_stages
+
+    def setup(self):
+        super(StagedCable, self).setup()
+
+        # Directories required by CABLE for outputs
+        os.makedirs(os.path.join(self.work_output_path, 'logs'),
+                    exist_ok=True)
+        os.makedirs(os.path.join(self.work_output_path, 'restart'),
+                    exist_ok=True)
+        os.makedirs(os.path.join(self.work_output_path, 'outputs'),
+                    exist_ok=True)
+
+        self._prepare_stage()
+
+    def _prepare_stage(self):
+        """Apply the stage namelist to the master namelist."""
+
+        if self.configuration_log['current_stage'] != '':
+            # If the current stage is a non-empty string, it means we exited
+            # during the running of the previous stage
+            stage_name = self.configuration_log['current_stage']
+        else:
+            # Pop the stage from the list
+            stage_name = self.configuration_log['queued_stages'].pop(0)
+            self.configuration_log['current_stage'] = stage_name
+
+        with open('configuration_log.yaml', 'w') as conf_log_f:
+            yaml.dump(self.configuration_log, conf_log_f)
+
+        # Ensure the directory exists
+        if not os.path.isdir(stage_name):
+            errmsg = f"""Directory containing namelists for stage {stage_name}
+             does not exist."""
+            raise FileNotFoundError(errmsg)
+
+        # Apply the stage namelists
+        self._apply_stage_namelists(stage_name)
+
+        # Copy the log to the work directory
+        shutil.copy('configuration_log.yaml', self.work_input_path)
+
+    def _apply_stage_namelists(self, stage_name):
+        """Apply the stage namelists to the master namelists."""
+        namelists = os.listdir(stage_name)
+
+        for namelist in namelists:
+            with open(namelist, 'r') as master_nml_f:
+                master_namelist = f90nml.read(master_nml_f)
+
+            with open(os.path.join(stage_name, namelist), 'r') as stage_nml_f:
+                stage_namelist = f90nml.read(stage_nml_f)
+
+            deep_update(master_namelist, stage_namelist)
```

Comment:
`f90nml` has a `patch` function that might do what you want

https://github.com/marshallward/f90nml/blob/21db6a7afc6b5df6e53cdc1980f1301660732d01/f90nml/namelist.py#L654-L668

The docs suggest you an call a higher level `patch` function without needing to do any reading of input and output namelists, just pass the filenames and the patch and it does the rest:
```
f90nml.patch('sample.nml', patch_nml, 'new_sample.nml')
```

Reply:
I tried this originally, but it wasn't handling struct attributes correctly. Turns out that the object method ```patch``` does not work as expected, but calling the library ```patch``` function as you recommend here does work as expected.